docs/specs/instrumentation.md

```diff
@@ -8,34 +8,148 @@ interpreted as described in [RFC 2119].
+This document strives to guide developers into achieving this.
 
 ## Naming
 
+### Event naming
+
+Vector implements an event-driven instrumentation pattern ([RFC 2064]) and
+event names MUST adhere to the following rules:
+
+* MUST only contain ASCII alphanumeric and lowercase characters
+* MUST be in [camelcase] format
+* MUST follow the `<Namespace><Noun><Verb>[Error]` template
+  * `Namespace` - the internal domain the event belongs to (e.g., `Component`, `Buffer`, `Topology`)
+  * `Noun` - the subject of the event (e.g., `Bytes`, `Events`)
+  * `Verb` - the past tense verb describing when the event occured (e.g., `Received`, `Sent`, `Processes`)
+  * `[Error]` - if the event is an error it MUST end with `Error`
+
 ### Metric naming
 
-For metric naming, Vector broadly follows the
-[Prometheus metric naming standards]. Hence, a metric name:
+Vector broadly follows the [Prometheus metric naming standards]:
 
-* MUST only contain ASCII alphanumeric, lowercase, and underscores
+* MUST only contain ASCII alphanumeric, lowercase, and underscore characters
+* MUST be in [snakecase] format
 * MUST follow the `<namespace>_<name>_<unit>_[total]` template
-  * `namespace` represents a broad category of metrics (e.g., `component`, `buffer`, `topology`)
-  * `name` is one or more words that describes the measurement (e.g., `memory_rss`, `requests`)
-  * `unit` MUST be a single [base unit] in plural form, if applicable (e.g., `seconds`, `bytes`)
+  * `namespace` - the internal domain that the metric belongs to (e.g., `component`, `buffer`, `topology`)
+  * `name` - is one or more words that describes the measurement (e.g., `memory_rss`, `requests`)
+  * `unit` - MUST be a single [base unit] in plural form, if applicable (e.g., `seconds`, `bytes`)
   * Counters MUST end with `total` (e.g., `disk_written_bytes_total`, `http_requests_total`)
 * SHOULD be broad in purpose and use use tags to differentiate characteristics of the measurement (e.g., `host_cpu_seconds_total{cpu="0",mode="idle"}`)
 
+## Emission
+
+### Batching
+
+For performance reasons, as demonstrated in [pull request #8383],
+instrumentation SHOULD be batched whenever possible:
+
+* Telemtry SHOULD emit for entire event batches, not each individual event.
+  [RFC 9480] describes Vector's batching strategy.
+* Benchmarking SHOULD prove that batching produces performance benefits.
+  [Issue 10658] could eliminate the need to batch for performance improvements.
+
+### Events
+
+Instrumentation SHOULD be event-driven ([RFC 2064]), where individual events
+serve as the vehicle for internal telemtry, driving the emission of metrics
+and logs. This organizes Vector's telemetry, making it easier to manage and 
+catalogue. On rare occassions, metrics and logs can emit directly but MUST be
+reserved for ocassions where it is impossible to emit Vector's events. For
+example, emitting metrics in a library that cannot import Vector's events.
```

Comment:
```suggestion
catalogue. Metrics and logs SHOULD NOT be emitted directly except for where it
is otherwise impossible to emit Vector's events, such as in an external crate
that cannot import Vector's events.